use "collidesWith" instead of "collisionMask"

AndrewMeadows opened the pull request:
...but make "collisionMask" work for those who want it

After thinking about it I decided that the string based "collisionMask" setting is a Bad Idea.  The problem is that it is not easy to do bitwise operations on strings.  This PR is a proposal for something else:

(1) Use "collidesWith" for people who want to use strings

(2) Reserve "collisionMask" for the correct way to do it

This is for you @zappoman and @imgntn

AndrewMeadows commented on examples/controllers/handControllerGrab.js:
```diff
@@ -1842,7 +1842,10 @@ function MyController(hand) {
+                //"collisionMask": COLLISION_MASK_WHILE_GRABBED | grabbedProperties.collisionMask
+                // when using string values
+                "collidesWith": COLLIDES_WITH_WHILE_GRABBED
             };
```
The problem here is that it is hard to preserve collision mask holes already set on the object when using strings.  Possible, but would require some JS helper functions to parse the string, use lookup tables to turn it into a numerical value, and then convert it back to a string.

ZappoMan replied:
It seems like this wouldn't be that hard to do in the EntityItemProperties class.

> On Jan 12, 2016, at 5:50 PM, Andrew Meadows <notifications@github.com> wrote:
> 
> In examples/controllers/handControllerGrab.js:
> 
> > @@ -1842,7 +1842,10 @@ function MyController(hand) {
> >                      y: 0,
> >                      z: 0
> >                  },
> > -                "collisionMask": COLLISION_MASK_WHILE_GRABBED
> > +                // bummer, it isn't easy to do bitwise collisionMask operations like this:
> > +                //"collisionMask": COLLISION_MASK_WHILE_GRABBED | grabbedProperties.collisionMask
> > +                // when using string values
> > +                "collidesWith": COLLIDES_WITH_WHILE_GRABBED
> >              };
> The problem here is that it is hard to preserve collision mask holes already set on the object when using strings. Possible, but would require some JS helper functions to parse the string, use lookup tables to turn it into a numerical value, and then convert it back to a string.
> 
> —
> Reply to this email directly or view it on GitHub.
>

AndrewMeadows replied:
In more detail:

The default behavior of the handControllerGrab.js script is to configure the grabbed object to collide only with: dynamic and otherAvatars.  This allows the grabber to poke dynamic objects and nearby avatars with the grabee, but to pass the grabbee through static table and walls.

However, suppose the object were already configured to NOT collide with dynamic objects and other avatars before being grabbed?  When the object becomes grabbed it will suddenly start colliding with these things when it before it did not.

What we'd like to do is combine the new mask and the old such that any overlapping groups are maintained and any conflicting groups are cleared.  This can be done by  ANDing the new mask with the old, and must be done in the JS context.  It is possible to compute the AND operation between lists of unique strings but it is inordinately complicated... because it is so very easy to do with numerical masks.

hifi-gustavo commented:
Build console output for commit: d5ebb144cf0dd031f2e738a4a2149cf8717e1426 available at: https://gist.github.com/7c157a6de091b4bfbf5a<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/pr-interface-win64-PR-6820.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6820.dmg

murillodigital commented:
test this please

hifi-gustavo commented:
Build console output for commit: d5ebb144cf0dd031f2e738a4a2149cf8717e1426 available at: https://gist.github.com/4d05c689a06151f039f9<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/pr-interface-win64-PR-6820.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6820.dmg

AndrewMeadows commented:
Unfortunately this work has already sat too long: another PR (#6830) has already used the **collisionMask** in the old style.  This PR is blocked by that one, and will have to be repaired once that other one is merge.

hifi-gustavo commented:
Build console output for commit: a11ff354640dccec39b9fbc2520c77ae57272676 available at: https://gist.github.com/4d05c689a06151f039f9<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/pr-interface-win64-PR-6820.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6820.dmg

AndrewMeadows commented:
This is ready to go.

nbq commented:
I tested with this bow script and this handgrab script and everything seemed to work ok for me.

AndrewMeadows commented:
This had a merge conflict but is ready to go again.

hifi-gustavo commented:
Build console output for commit: 9c9ffd01f0c1c55f9a26ad559475e78260fc1d92 available at: https://gist.github.com/4d05c689a06151f039f9<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/pr-interface-win64-PR-6820.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6820.dmg

hifi-gustavo commented:
Build console output for commit: bd751d289eb14f19c878e81c8be8ddec5f4f23ee available at: https://gist.github.com/4d05c689a06151f039f9<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/pr-interface-win64-PR-6820.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6820.dmg

ZappoMan commented:
@AndrewMeadows - this PR has gotten stale and needs some attention... how do you want to proceed? from a Code Review perspective, I am happy to merge it, once it has been updated or rebased... has someone else tested it? If not, how do you recommend testing it?

hifi-gustavo commented:
Build console output for commit: 839eeb6b8a0f0895c5de23f12d0b199687e7722e available at: https://gist.github.com/4d05c689a06151f039f9<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/pr-interface-win64-PR-6820.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6820.dmg

AndrewMeadows commented:
Yup, I just looked and realized that I had conflicts with my other recent PRs.  I rebased and fixed.

@nbq tested the **bow** and **handControllerGrab** scripts already.

There is an untested feature in this PR (well, tested by me before I submitted it but by no one else as far as I know):  the "collisionMask" property can be used (instead of "collidesWith") to perform bitwise operations on collision masks which will allow me to make the **handControllerGrab** script have more correct behavior -- I didn't want to modify it as part of this PR because I wanted all tests to be on the "collidesWith" property.

Go ahead and merge.  After that I'll update **handControllerGrab**,  submit as its own PR, and write up a  test plan for it.

hifi-gustavo commented:
Build console output for commit: 4a91ed3d6632ff7927d6ec0651353f39f49fb493 available at: https://gist.github.com/4d05c689a06151f039f9<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/pr-interface-win64-PR-6820.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6820.dmg
